File: src/benchmark.cpp
#include "blocklab/Agent.h"
#include "blocklab/CliParsing.h"
#include "blocklab/Environment.h"
#include "blocklab/Renderer.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <memory>
#include <random>
#include <string_view>
#include <vector>

namespace {

struct BenchmarkConfig {
    uint64_t steps = 0;
    uint64_t warmupSteps = 128;
    double seconds = 10.0;
    double reportInterval = 1.0;
    uint32_t seed = 1;
    uint32_t batchSize = 1;
    int32_t worldRadiusChunks = 4;
    bool visualize = false;
    double visualFps = 30.0;
    int32_t minActionSteps = 20;
    int32_t maxActionSteps = 160;
    uint64_t initialOverrides = 1000;
    blocklab::RenderConfig renderConfig { .width = 320, .height = 180, .visible = false, .present = false };
    blocklab::RenderConfig visualConfig { .width = 320, .height = 180, .visible = true, .present = true };
};

struct BenchmarkStats {
    uint64_t iterations = 0;
    uint64_t episodes = 0;
    uint64_t blocksCollected = 0;
    uint64_t blocksPlaced = 0;
    double reward = 0.0;
};

class RandomAgent {
public:
    explicit RandomAgent(const BenchmarkConfig& config)
        : m_minActionSteps(config.minActionSteps)
        , m_maxActionSteps(config.maxActionSteps)
    {
    }

    blocklab::AgentAction nextAction(std::mt19937& rng)
    {
        if (m_remainingSteps == 0) {
            std::uniform_int_distribution<int32_t> duration(m_minActionSteps, m_maxActionSteps);
            std::uniform_real_distribution<float> walk(0.45f, 1.0f);
            std::uniform_real_distribution<float> strafe(-0.35f, 0.35f);
            std::normal_distribution<float> turn(0.0f, 0.012f);
            std::bernoulli_distribution walkForward(0.82);
            m_remainingSteps = duration(rng);
            m_action = {};
            m_action.forward = walkForward(rng) ? walk(rng) : 0.0f;
            m_action.right = strafe(rng);
            m_action.yawDelta = turn(rng);
            m_action.pitchDelta = turn(rng) * 0.25f;
        }

        blocklab::AgentAction action = m_action;
        std::bernoulli_distribution jump(0.004);
        std::bernoulli_distribution dig(0.004);
        std::bernoulli_distribution place(0.003);
        action.jump = jump(rng);
        action.dig = dig(rng);
        action.place = !action.dig && place(rng);
        --m_remainingSteps;
        return action;
    }

private:
    int32_t m_minActionSteps = 20;
    int32_t m_maxActionSteps = 160;
    int32_t m_remainingSteps = 0;
    blocklab::AgentAction m_action;
};

[[noreturn]] void usage(int exitCode)
{
    std::printf("Usage: blocklab_benchmark [options]\n"
                "\n"
                "Options:\n"
                "  --seconds N              Run for N seconds, default 10. Ignored when --steps is non-zero.\n"
                "  --steps N                Run fixed number of benchmark iterations, default 0.\n"
                "  --warmup-steps N         Run N untimed warmup iterations before measurement, default 128.\n"
                "  --batch-size N           Number of environments stepped and rendered per iteration, default 1.\n"
                "  --num-envs N             Alias for --batch-size.\n"
                "  --seed N                 RNG seed, default 1.\n"
                "  --world-radius N         World radius in chunks, default 4.\n"
                "  --report-interval N      Progress report interval in seconds, default 1.\n"
                "  --visualize              Open a Vulkan window and render the environment visibly.\n"
                "  --visual-fps N           Max visible presentation rate, default 30.\n"
                "  --action-steps A:B       Hold sampled movement/look for A..B steps, default 20:160.\n"
                "  --initial-overrides N    Apply N clustered block overrides after each reset, default 1000.\n"
                "  --resolution WxH         Render/window resolution, default 320x180.\n"
                "  -h, --help               Show this help.\n");
    std::exit(exitCode);
}

BenchmarkConfig parseArgs(int argc, char** argv)
{
    BenchmarkConfig config;
    for (int i = 1; i < argc; ++i) {
        const std::string_view arg(argv[i]);
        if (arg == "-h" || arg == "--help")
            usage(EXIT_SUCCESS);
        else if (arg == "--visualize")
            config.visualize = true;
        else if (arg == "--seconds" || arg.starts_with("--seconds="))
            config.seconds = blocklab::cli::parseDouble(blocklab::cli::optionValue(i, argc, argv, arg, "--seconds"))
                                 .value_or(config.seconds);
        else if (arg == "--steps" || arg.starts_with("--steps="))
            config.steps = blocklab::cli::parseInt<uint64_t>(blocklab::cli::optionValue(i, argc, argv, arg, "--steps"))
                               .value_or(config.steps);
        else if (arg == "--warmup-steps" || arg.starts_with("--warmup-steps="))
            config.warmupSteps
                = blocklab::cli::parseInt<uint64_t>(blocklab::cli::optionValue(i, argc, argv, arg, "--warmup-steps"))
                      .value_or(config.warmupSteps);
        else if (arg == "--batch-size" || arg.starts_with("--batch-size=") || arg == "--num-envs"
            || arg.starts_with("--num-envs=")) {
            const char* const optionName
                = (arg == "--num-envs" || arg.starts_with("--num-envs=")) ? "--num-envs" : "--batch-size";
            const auto batchSize
                = blocklab::cli::parseInt<uint64_t>(blocklab::cli::optionValue(i, argc, argv, arg, optionName));
            if (!batchSize || *batchSize == 0 || *batchSize > static_cast<uint64_t>(UINT32_MAX)) [[unlikely]] {
                std::fprintf(stderr, "Invalid %s value.\n", optionName);
                std::exit(EXIT_FAILURE);
            }
            config.batchSize = static_cast<uint32_t>(*batchSize);
        } else if (arg == "--seed" || arg.starts_with("--seed="))
            config.seed = static_cast<uint32_t>(
                blocklab::cli::parseInt<uint64_t>(blocklab::cli::optionValue(i, argc, argv, arg, "--seed"))
                    .value_or(config.seed));
        else if (arg == "--world-radius" || arg.starts_with("--world-radius=")) {
            const auto worldRadiusChunks
                = blocklab::cli::parseInt<int32_t>(blocklab::cli::optionValue(i, argc, argv, arg, "--world-radius"));
            if (!worldRadiusChunks || *worldRadiusChunks <= 0) [[unlikely]] {
                std::fprintf(stderr, "Invalid --world-radius value.\n");
                std::exit(EXIT_FAILURE);
            }
            config.worldRadiusChunks = *worldRadiusChunks;
        } else if (arg == "--report-interval" || arg.starts_with("--report-interval=")) {
            const auto reportInterval
                = blocklab::cli::parseDouble(blocklab::cli::optionValue(i, argc, argv, arg, "--report-interval"));
            if (!reportInterval || *reportInterval <= 0.0) [[unlikely]] {
                std::fprintf(stderr, "Invalid --report-interval value.\n");
                std::exit(EXIT_FAILURE);
            }
            config.reportInterval = *reportInterval;
        } else if (arg == "--visual-fps" || arg.starts_with("--visual-fps=")) {
            const auto visualFps
                = blocklab::cli::parseDouble(blocklab::cli::optionValue(i, argc, argv, arg, "--visual-fps"));
            if (!visualFps || *visualFps <= 0.0) [[unlikely]] {
                std::fprintf(stderr, "Invalid --visual-fps value.\n");
                std::exit(EXIT_FAILURE);
            }
            config.visualFps = *visualFps;
        } else if (arg == "--action-steps" || arg.starts_with("--action-steps=")) {
            const auto actionSteps
                = blocklab::cli::parseActionSteps(blocklab::cli::optionValue(i, argc, argv, arg, "--action-steps"));
            if (!actionSteps) [[unlikely]] {
                std::fprintf(stderr, "Invalid --action-steps value. Expected MIN:MAX.\n");
                std::exit(EXIT_FAILURE);
            }
            config.minActionSteps = actionSteps->first;
            config.maxActionSteps = actionSteps->second;
        } else if (arg == "--resolution" || arg.starts_with("--resolution=")) {
            auto renderConfig
                = blocklab::cli::parseResolution(blocklab::cli::optionValue(i, argc, argv, arg, "--resolution"));
            if (!renderConfig) [[unlikely]] {
                std::fprintf(stderr, "Invalid --resolution value. Expected WIDTHxHEIGHT.\n");
                std::exit(EXIT_FAILURE);
            }
            renderConfig->visible = config.renderConfig.visible;
            renderConfig->present = config.renderConfig.present;
            config.renderConfig = *renderConfig;
            config.visualConfig.width = renderConfig->width;
            config.visualConfig.height = renderConfig->height;
        } else if (arg == "--initial-overrides" || arg.starts_with("--initial-overrides="))
            config.initialOverrides = blocklab::cli::parseInt<uint64_t>(
                blocklab::cli::optionValue(i, argc, argv, arg, "--initial-overrides"))
                                          .value_or(config.initialOverrides);
        else
            usage(EXIT_FAILURE);
    }
    config.renderConfig.batchSize = config.batchSize;
    config.visualConfig.batchSize = 1;
    return config;
}

uint64_t applyInitialOverrides(blocklab::World& world, const BenchmarkConfig& config, uint32_t seed)
{
    const std::size_t before = world.overrideCount();
    std::mt19937 rng(seed ^ 0x8f3d5b79U);
    std::normal_distribution<float> horizontal(0.0f, 8.0f);
    std::uniform_int_distribution<int32_t> verticalOffset(-8, 4);
    for (uint64_t attempts = 0;
         world.overrideCount() - before < config.initialOverrides && attempts < config.initialOverrides * 64 + 1024;
         ++attempts) {
        const int32_t x = std::clamp(static_cast<int32_t>(std::lround(horizontal(rng))), -28, 28);
        const int32_t z = std::clamp(static_cast<int32_t>(std::lround(horizontal(rng))), -28, 28);
        const int32_t surfaceY
            = std::max(0, blocklab::floorToInt32(world.groundHeight(static_cast<float>(x), static_cast<float>(z))) - 1);
        const int32_t y = std::clamp(surfaceY + verticalOffset(rng), 0, blocklab::Chunk::SizeY - 1);
        const blocklab::Block current = world.getBlock(x, y, z);
        world.setBlock(x, y, z, current == blocklab::Block::Air ? blocklab::Block::Stone : blocklab::Block::Air);
    }
    return static_cast<uint64_t>(world.overrideCount() - before);
}

uint64_t resetEnvironment(blocklab::Environment& env, const BenchmarkConfig& config, uint32_t seed)
{
    env.reset(seed);
    uint64_t applied = 0;
    for (uint32_t i = 0; i < env.batchSize(); ++i)
        applied += applyInitialOverrides(env.mutableWorld(i), config, seed + i * 0x9e3779b9U);
    return applied;
}

} // namespace

int main(int argc, char** argv)
{
    const BenchmarkConfig config = parseArgs(argc, argv);
    std::mt19937 rng(config.seed);
    std::vector<RandomAgent> randomAgents;
    randomAgents.reserve(config.batchSize);
    for (uint32_t i = 0; i < config.batchSize; ++i)
        randomAgents.emplace_back(config);
    std::vector<blocklab::AgentAction> actions(config.batchSize);
    blocklab::Renderer renderer(config.renderConfig);
    blocklab::Environment env(renderer, config.batchSize, config.worldRadiusChunks);
    std::unique_ptr<blocklab::Renderer> visualRenderer;
    if (config.visualize)
        visualRenderer = std::make_unique<blocklab::Renderer>(config.visualConfig);
    uint64_t lastInitialOverridesApplied = resetEnvironment(env, config, config.seed);

    for (uint64_t step = 0; step < config.warmupSteps; ++step) {
        renderer.pollEvents();
        for (uint32_t i = 0; i < config.batchSize; ++i)
            actions[i] = randomAgents[i].nextAction(rng);
        const std::span<const blocklab::StepResult> results = env.step(actions);
        const bool resetNeeded = std::any_of(results.begin(), results.end(),
            [](const blocklab::StepResult& result) { return result.terminated || result.truncated; });
        if (resetNeeded)
            lastInitialOverridesApplied = resetEnvironment(env, config, config.seed + static_cast<uint32_t>(step + 1));
    }

    using Clock = std::chrono::steady_clock;
    const auto startedAt = Clock::now();
    auto lastReportAt = startedAt;
    const auto visualInterval
        = std::chrono::duration_cast<Clock::duration>(std::chrono::duration<double>(1.0 / config.visualFps));
    auto lastVisualAt = startedAt - visualInterval;
    uint64_t lastReportIterations = 0;
    BenchmarkStats stats;
    std::vector<int32_t> lastBlocksCollected(config.batchSize);
    std::vector<int32_t> lastBlocksPlaced(config.batchSize);
    for (uint32_t i = 0; i < config.batchSize; ++i) {
        lastBlocksCollected[i] = env.agent(i).state().blocksCollected;
        lastBlocksPlaced[i] = env.agent(i).state().blocksPlaced;
    }

    while (true) {
        renderer.pollEvents();

        if (visualRenderer) {
            visualRenderer->pollEvents();
            if (visualRenderer->shouldClose())
                break;
        }
        for (uint32_t i = 0; i < config.batchSize; ++i)
            actions[i] = randomAgents[i].nextAction(rng);
        const std::span<const blocklab::StepResult> results = env.step(actions);
        ++stats.iterations;
        bool resetNeeded = false;
        for (uint32_t i = 0; i < config.batchSize; ++i) {
            const blocklab::StepResult& result = results[i];
            stats.reward += result.reward;
            const blocklab::AgentState& agent = env.agent(i).state();
            stats.blocksCollected += static_cast<uint64_t>(std::max(0, agent.blocksCollected - lastBlocksCollected[i]));
            stats.blocksPlaced += static_cast<uint64_t>(std::max(0, agent.blocksPlaced - lastBlocksPlaced[i]));
            lastBlocksCollected[i] = agent.blocksCollected;
            lastBlocksPlaced[i] = agent.blocksPlaced;
            if (result.terminated || result.truncated) {
                ++stats.episodes;
                resetNeeded = true;
            }
        }

        if (resetNeeded) {
            lastInitialOverridesApplied
                = resetEnvironment(env, config, config.seed + static_cast<uint32_t>(stats.episodes));
            for (uint32_t i = 0; i < config.batchSize; ++i) {
                lastBlocksCollected[i] = env.agent(i).state().blocksCollected;
                lastBlocksPlaced[i] = env.agent(i).state().blocksPlaced;
            }
        }

        const auto now = Clock::now();
        if (visualRenderer && now - lastVisualAt >= visualInterval) {
            const blocklab::AgentState agents[] = { env.agent(0).state() };
            visualRenderer->renderObservations({ &env.world(0), 1 }, agents);
            lastVisualAt = now;
        }
        const double elapsed = std::chrono::duration<double>(now - startedAt).count();
        if (std::chrono::duration<double>(now - lastReportAt).count() >= config.reportInterval) {
            const double interval = std::chrono::duration<double>(now - lastReportAt).count();
            const uint64_t intervalIterations = stats.iterations - lastReportIterations;
            const uint64_t totalSteps = stats.iterations * config.batchSize;
            const double intervalStepsPerSecond = static_cast<double>(intervalIterations * config.batchSize) / interval;
            std::printf("iterations=%llu steps=%llu elapsed=%.2fs steps/s=%.0f avg_reward=%.4f episodes=%llu\n",
                static_cast<unsigned long long>(stats.iterations), static_cast<unsigned long long>(totalSteps), elapsed,
                intervalStepsPerSecond, stats.reward / static_cast<double>(std::max<uint64_t>(1, totalSteps)),
                static_cast<unsigned long long>(stats.episodes));
            lastReportAt = now;
            lastReportIterations = stats.iterations;
        }
        if ((config.steps > 0 && stats.iterations >= config.steps) || (config.steps == 0 && elapsed >= config.seconds))
            break;
    }

    const double totalSeconds = std::chrono::duration<double>(Clock::now() - startedAt).count();
    const blocklab::Observation& observation = env.observe();
    uint64_t worldOverrides = 0;
    for (uint32_t i = 0; i < config.batchSize; ++i)
        worldOverrides += env.world(i).overrideCount();
    const uint64_t totalSteps = stats.iterations * config.batchSize;
    std::printf("\nBlockLab benchmark result\n"
                "  batch_size: %u\n"
                "  iterations: %llu\n"
                "  total_steps: %llu\n"
                "  warmup_steps: %llu\n"
                "  total_time_s: %.4f\n"
                "  steps_per_second: %.2f\n"
                "  avg_reward: %.6f\n"
                "  episodes: %llu\n"
                "  blocks_collected: %llu\n"
                "  blocks_placed: %llu\n"
                "  initial_overrides_requested: %llu\n"
                "  initial_overrides_applied: %llu\n"
                "  world_overrides: %llu\n"
                "  observation_version: %llu\n"
                "  observation_handle: 0x%llx\n",
        config.batchSize, static_cast<unsigned long long>(stats.iterations),
        static_cast<unsigned long long>(totalSteps), static_cast<unsigned long long>(config.warmupSteps), totalSeconds,
        static_cast<double>(totalSteps) / totalSeconds,
        stats.reward / static_cast<double>(std::max<uint64_t>(1, totalSteps)),
        static_cast<unsigned long long>(stats.episodes), static_cast<unsigned long long>(stats.blocksCollected),
        static_cast<unsigned long long>(stats.blocksPlaced), static_cast<unsigned long long>(config.initialOverrides),
        static_cast<unsigned long long>(lastInitialOverridesApplied), static_cast<unsigned long long>(worldOverrides),
        static_cast<unsigned long long>(observation.version()), static_cast<unsigned long long>(observation.handle(0)));
}

File: python/blocklab/examples/benchmark.py
from __future__ import annotations

import argparse
import random
import time

import torch

import blocklab


DEFAULT_MAX_STEPS = 60 * 60 * 80  # 4 game days.


class RandomAgent:
    def __init__(self, *, min_action_steps: int, max_action_steps: int) -> None:
        self._min_action_steps = min_action_steps
        self._max_action_steps = max_action_steps
        self._remaining_steps = 0
        self._forward = 0.0
        self._right = 0.0
        self._yaw_delta = 0.0
        self._pitch_delta = 0.0

    def next_action(self, rng: random.Random):
        if self._remaining_steps == 0:
            self._remaining_steps = rng.randint(self._min_action_steps, self._max_action_steps)
            self._forward = rng.uniform(0.45, 1.0) if rng.random() < 0.82 else 0.0
            self._right = rng.uniform(-0.35, 0.35)
            self._yaw_delta = rng.gauss(0.0, 0.012)
            self._pitch_delta = rng.gauss(0.0, 0.012) * 0.25

        action = blocklab.AgentAction()
        action.forward = self._forward
        action.right = self._right
        action.yaw_delta = self._yaw_delta
        action.pitch_delta = self._pitch_delta
        action.jump = rng.random() < 0.004
        action.attack = rng.random() < 0.004
        action.use = (not action.attack) and rng.random() < 0.003
        self._remaining_steps -= 1
        return action


def parse_action_steps(value: str) -> tuple[int, int]:
    parts = value.split(":", maxsplit=1)
    if len(parts) != 2:
        raise argparse.ArgumentTypeError("expected MIN:MAX")
    min_steps = int(parts[0])
    max_steps = int(parts[1])
    if min_steps <= 0 or max_steps < min_steps:
        raise argparse.ArgumentTypeError("expected 0 < MIN <= MAX")
    return min_steps, max_steps


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--num-envs", type=int, default=1)
    parser.add_argument("--steps", type=int, default=0)
    parser.add_argument("--seconds", type=float, default=10.0)
    parser.add_argument("--width", type=int, default=160)
    parser.add_argument("--height", type=int, default=90)
    parser.add_argument("--device", default=None)
    parser.add_argument("--warmup-steps", type=int, default=128)
    parser.add_argument("--report-interval", type=float, default=1.0)
    parser.add_argument("--action-steps", type=parse_action_steps, default=(20, 160))
    parser.add_argument("--seed", type=int, default=1)
    parser.add_argument("--max-steps", type=int, default=DEFAULT_MAX_STEPS)
    parser.add_argument("--materialize-observation", action="store_true")
    args = parser.parse_args()

    env = blocklab.make(
        num_envs=args.num_envs,
        width=args.width,
        height=args.height,
        device=args.device,
        seed=args.seed,
        max_steps=args.max_steps,
    )
    rng = random.Random(args.seed)
    random_agents = [
        RandomAgent(min_action_steps=args.action_steps[0], max_action_steps=args.action_steps[1])
        for _ in range(args.num_envs)
    ]
    obs, _ = env.reset()
    tensor = torch.from_dlpack(obs) if args.materialize_observation else None

    for step in range(args.warmup_steps):
        actions = [agent.next_action(rng) for agent in random_agents]
        obs, reward, terminated, truncated, info = env.step(actions)
        if args.materialize_observation:
            tensor = torch.from_dlpack(obs)
        if any(done or cut for done, cut in zip(info["terminated"], info["truncated"])):
            obs, _ = env.reset(seed=args.seed + step + 1)
            if args.materialize_observation:
                tensor = torch.from_dlpack(obs)

    started = time.perf_counter()
    last_report_at = started
    last_report_step = 0
    total_reward = 0.0
    total_episodes = 0
    step = 0
    while True:
        step += 1
        actions = [agent.next_action(rng) for agent in random_agents]
        obs, reward, terminated, truncated, info = env.step(actions)
        if args.materialize_observation:
            tensor = torch.from_dlpack(obs)
        total_reward += float(sum(info["reward"]))
        episode_count = sum(done or cut for done, cut in zip(info["terminated"], info["truncated"]))
        total_episodes += episode_count
        if episode_count > 0:
            obs, _ = env.reset(seed=args.seed + total_episodes)
            if args.materialize_observation:
                tensor = torch.from_dlpack(obs)
        now = time.perf_counter()
        if args.report_interval > 0.0 and now - last_report_at >= args.report_interval:
            interval = now - last_report_at
            interval_steps = (step - last_report_step) * args.num_envs
            total_steps = step * args.num_envs
            avg_reward = total_reward / max(1, total_steps)
            print(
                f"iterations={step} steps={total_steps} elapsed={now - started:.2f}s "
                f"steps/s={interval_steps / interval:.0f} avg_reward={avg_reward:.4f} episodes={total_episodes}",
                flush=True,
            )
            last_report_at = now
            last_report_step = step
        if args.steps > 0:
            if step >= args.steps:
                break
        elif now - started >= args.seconds:
            break
    elapsed = time.perf_counter() - started
    total_steps = step * args.num_envs

    print(f"device: {env.observation_spec.device}")
    print(f"observation_shape: {env.observation_spec.shape}")
    print(f"observation_dtype: {obs.dtype}")
    print(f"materialize_observation: {args.materialize_observation}")
    if tensor is not None:
        print(f"materialized_tensor_ptr: {tensor.data_ptr()}")
    print(f"warmup_steps: {args.warmup_steps}")
    print(f"iterations: {step}")
    print(f"total_steps: {total_steps}")
    print(f"avg_reward: {total_reward / max(1, total_steps):.6f}")
    print(f"episodes: {total_episodes}")
    print(f"steps_per_second: {total_steps / elapsed:.0f}")


if __name__ == "__main__":
    main()
